File: GKPageScrollViewObjc/GKPageScrollViewObjc/Demo/Nest/GKNestScrollView.m
//
//  GKNestScrollView.m
//  GKPageScrollViewObjc
//
//  Created by gaokun on 2019/10/21.
//  Copyright © 2019 gaokun. All rights reserved.
//

#import "GKNestScrollView.h"

@implementation GKNestScrollView

- (BOOL)gestureRecognizerShouldBegin:(UIGestureRecognizer *)gestureRecognizer {
    if ([self.gestureDelegate respondsToSelector:@selector(nestScrollView:gestureRecognizerShouldBegin:)]) {
        return [self.gestureDelegate nestScrollView:self gestureRecognizerShouldBegin:gestureRecognizer];
    }
    return YES;
}

// 允许多个手势
- (BOOL)gestureRecognizer:(UIGestureRecognizer *)gestureRecognizer shouldRecognizeSimultaneouslyWithGestureRecognizer:(UIGestureRecognizer *)otherGestureRecognizer {
    if ([self.gestureDelegate respondsToSelector:@selector(nestScrollView:gestureRecognizer:shouldRecognizeSimultaneouslyWithGestureRecognizer:)]) {
        return [self.gestureDelegate nestScrollView:self gestureRecognizer:gestureRecognizer shouldRecognizeSimultaneouslyWithGestureRecognizer:otherGestureRecognizer];
    }
    return NO;
}

@end

File: GKPageScrollViewSwift/GKPageScrollViewSwift/Demo/Nest/GKNestScrollView.swift
//
//  GKNestScrollView.swift
//  GKPageScrollViewSwift
//
//  Created by gaokun on 2019/10/22.
//  Copyright © 2019 gaokun. All rights reserved.
//

import UIKit

protocol GKNestScrollViewDelegate: AnyObject {
    func nestScrollViewGestureRecognizerShouldBegin(_ scrollView: GKNestScrollView, gestureRecognizer: UIGestureRecognizer) -> Bool
    func nestScrollGestureRecognizer(_ scrollView: GKNestScrollView, gestureRecognizer: UIGestureRecognizer, shouldRecognizeSimultaneouslyWith otherGestureRecognizer: UIGestureRecognizer) -> Bool
}

class GKNestScrollView: UIScrollView {

    public weak var gestureDelegate: GKNestScrollViewDelegate?
    
    override func gestureRecognizerShouldBegin(_ gestureRecognizer: UIGestureRecognizer) -> Bool {
        if let result = self.gestureDelegate?.nestScrollViewGestureRecognizerShouldBegin(self, gestureRecognizer: gestureRecognizer) {
            return result
        }
        
        return true
    }

    func gestureRecognizer(_ gestureRecognizer: UIGestureRecognizer, shouldRecognizeSimultaneouslyWith otherGestureRecognizer: UIGestureRecognizer) -> Bool {
        if let result = self.gestureDelegate?.nestScrollGestureRecognizer(self, gestureRecognizer: gestureRecognizer, shouldRecognizeSimultaneouslyWith: otherGestureRecognizer) {
            return result
        }
        
        return false
    }
}
